vpr/src/route/rr_graph_uxsdcxx_interface_impl.h

```diff
@@ -0,0 +1,1729 @@
+                              nullptr);
+    }
+    inline bool has_edge_metadata(const void* /*data*/, void* /*iter*/) override {
+        return rr_edge_metadata_.find(
+                   std::make_tuple(
+                       walker_.current_src_node(),
+                       walker_.current_sink_node(),
+                       walker_.current_switch_id_node()))
+               != rr_edge_metadata_.end();
+    }
+    inline size_t num_rr_edges_edge(const void* /*data*/, void* /*iter*/) override {
+        return walker_.initialize(rr_nodes_);
+    }
+    inline std::pair<const void*, void*> get_rr_edges_edge(int n, const void* /*data*/, void* /*iter*/) override {
+        VTR_ASSERT((ssize_t)walker_.advance() == n);
+        return std::make_pair(nullptr, nullptr);
+    }
+
+    /** Generated for complex type "node_timing":
+     * <xs:complexType name="node_timing">
+     *   <xs:attribute name="R" type="xs:float" use="required" />
+     *   <xs:attribute name="C" type="xs:float" use="required" />
+     * </xs:complexType>
+     */
+
+    /** Generated for complex type "node_segment":
+     * <xs:complexType name="node_segment">
+     *   <xs:attribute name="segment_id" type="xs:int" use="required" />
+     * </xs:complexType>
+     */
+
+    /** Generated for complex type "node":
+     * <xs:complexType name="node">
+     *   <xs:all>
+     *     <xs:element name="loc" type="node_loc" />
+     *     <xs:element minOccurs="0" name="timing" type="node_timing" />
+     *     <xs:element minOccurs="0" name="segment" type="node_segment" />
+     *     <xs:element minOccurs="0" name="metadata" type="metadata" />
+     *   </xs:all>
+     *   <xs:attribute name="id" type="xs:unsignedInt" use="required" />
+     *   <xs:attribute name="type" type="node_type" use="required" />
+     *   <xs:attribute name="direction" type="node_direction" />
+     *   <xs:attribute name="capacity" type="xs:unsignedInt" use="required" />
+     * </xs:complexType>
+     */
+    inline void set_node_direction(uxsd::enum_node_direction direction, const void* data, void* /*iter*/) override {
+        const int* inode = static_cast<const int*>(data);
+        auto& node = (*rr_nodes_)[*inode];
+        switch (direction) {
+            case uxsd::enum_node_direction::UXSD_INVALID:
+                VTR_ASSERT(!(node.type() == CHANX || node.type() == CHANY));
+                return;
+            case uxsd::enum_node_direction::INC_DIR:
+                node.set_direction(INC_DIRECTION);
+                break;
+            case uxsd::enum_node_direction::DEC_DIR:
+                node.set_direction(DEC_DIRECTION);
+                break;
+            case uxsd::enum_node_direction::BI_DIR:
+                node.set_direction(BI_DIRECTION);
+                break;
+            default:
+                VPR_FATAL_ERROR(VPR_ERROR_OTHER,
+                                "Invalid node direction %d", direction);
+        }
+    }
```

Comment:
I guess you are suggesting that the convert from uxsd enum to the VPR enum should be in a function, and this method becomes: `{ node.set_direction(conv_direction(direction)); } `?

Reply:
> I guess you are suggesting that the conversion from uxsd enum to the VPR enum should be in a function

Something like that would help. However more broadly, my main concern is that the current code is difficult to follow since it mixes lots of details related to the parser (e.g. uxsd types) and the actual RR graph data structure construction.

Reply:
Take a look now, between the context/interface refactoring and the moving of the enums, I think I've addressed this issue.